File: data/scripts/kmer.py
#http://rosalind.info/problems/kmer/

from itertools import product

K = 4
ALFABETO = "ACGT"

def lettura(filename):
    data = []
    with open(filename) as f:
        record = ""
        first = True
        for riga in f:
            riga = riga.rstrip("\n")
            if riga[0] == ">":
                if not first:
                    data.append(record)
                record = ""
                first = False
            else:
                record += riga
    data.append(record)
    return data

def conta_kmer(s, k):
    conteggi = {}
    for i in range(len(s) - k + 1):
        kmer = s[i:i + k]
        conteggi[kmer] = conteggi.get(kmer, 0) + 1
    return conteggi

def main():
    seqs = lettura("dataset.txt")
    s = seqs[0]
    conteggi = conta_kmer(s, K)
    kmers_ordinati = ["".join(t) for t in product(ALFABETO, repeat=K)]
    risultato = [str(conteggi.get(kmer, 0)) for kmer in kmers_ordinati]
    print(" ".join(risultato))

if __name__ == "__main__":
    # execute only if run as a script
    main()

File: public/js-solutions/kmer.mjs
// k-Mer Composition (Rosalind ID: KMER) - soluzione JavaScript
// indipendente, non una trascrizione di problem.py: stessa logica
// (parsing FASTA, conteggio di tutti i 4-mer nella sequenza, poi
// output nell'ordine lessicografico dei 4^4 possibili 4-mer
// sull'alfabeto ACGT), riscritta in modo idiomatico per JS.
//
// L'ordine dei k-mer riproduce quello di
// itertools.product("ACGT", repeat=4): l'ultima posizione varia più
// velocemente, esattamente come in lexf.mjs.
//
// Contratto: riceve il contenuto testuale di dataset.txt, restituisce
// l'output testuale (stessa forma dell'output Python).
const K = 4;
const ALFABETO = "ACGT";

function parseFasta(testo) {
  const righe = testo.split("\n").map((r) => r.replace(/\r$/, ""));
  const data = [];
  let record = "";
  let first = true;
  for (const riga of righe) {
    if (riga.startsWith(">")) {
      if (!first) data.push(record);
      record = "";
      first = false;
    } else {
      record += riga;
    }
  }
  data.push(record);
  return data;
}

function contaKmer(s, k) {
  const conteggi = new Map();
  for (let i = 0; i <= s.length - k; i++) {
    const kmer = s.slice(i, i + k);
    conteggi.set(kmer, (conteggi.get(kmer) ?? 0) + 1);
  }
  return conteggi;
}

function* prodotto(simboli, k) {
  if (k === 0) {
    yield "";
    return;
  }
  for (const s of simboli) {
    for (const resto of prodotto(simboli, k - 1)) {
      yield s + resto;
    }
  }
}

export default function solve(datasetText) {
  const seqs = parseFasta(datasetText);
  const s = seqs[0] ?? "";

  if (!s) {
    throw new Error("Input non valido: nessuna sequenza FASTA trovata");
  }

  const conteggi = contaKmer(s, K);
  const risultato = [];
  for (const kmer of prodotto(ALFABETO, K)) {
    risultato.push(conteggi.get(kmer) ?? 0);
  }

  return `${risultato.join(" ")}\n`;
}
